lucene/analysis/common/src/java/org/apache/lucene/analysis/hunspell/Dictionary.java

```diff
@@ -1571,14 +1549,28 @@ boolean hasFlag(int entryId, char flag) {
     return flagLookup.hasFlag(entryId, flag);
   }
 
-  CharSequence cleanInput(CharSequence input, StringBuilder reuse) {
-    return cleanInput(input, input.length(), reuse);
+  boolean mayNeedInputCleaning() {
+    return ignoreCase || ignore != null || iconv != null;
+  }
+
+  boolean needsInputCleaning(CharSequence input) {
+    if (mayNeedInputCleaning()) {
+      for (int i = 0; i < input.length(); i++) {
+        char ch = input.charAt(i);
+        if (ignore != null && Arrays.binarySearch(ignore, ch) >= 0
+            || ignoreCase && caseFold(ch) != ch
+            || iconv != null && iconv.mightReplaceChar(ch)) {
```

Comment:
Ideally `cleanInput` would check all this and just return its input if there's nothing to change, but that's not terribly trivial if we want to avoid allocations with its `StringBuilder reuse` parameter. Maybe later.